Autorepeatable keybindings & keybindings for label brush size

RobAnKo opened the pull request:
# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
As discussed in #3994, a keybinding to allow changing brush sizes in a Label layer is a helpful feature for manual labeling.
In addition, allowing the keybinding to autorepeat (hold the key to continuously increase/decrease the brush size) makes life even easier.
Upon implementing a hacky, hard-coded solution in PR #5024, we realized that a general pattern for autorepeatable keybindings would be a useful feature. Here I present a working solution.

<!-- Tell us about your new feature, improvement, or fix! -->
Based on suggestions by @Czaki, 
- the Action class now has an additional field `repeatable : bool`, a flag that defines for each action whether it is autorepeatable.
- the ActionManager class has new methods `_is_repeatable` to check whether an action is repeatable, as well as `_get_repeatable_shortcuts` that returns a list of repeatable shortcuts. 
- At the moment the repeatability of actions is stored in a {name:bool} dictionary at napari/utils/repeatable_actions.py with the same keys as napari/utils/shortcuts.py
- In napari/utils/key_bindings  - on_key_press(), a call to action_manager._get_repeatable_shortcuts() is used to distinguish for a triggered action whether it is repeatable or not. 


Here you can see that press-and-hold works and that changing the hotkeys (from Q,E to Alt+A,Alt+D) in the settings does not break that functionality.
![output_reduced](https://user-images.githubusercontent.com/36212786/190443205-02ea8687-4ce5-4d66-a87a-c406692635c9.gif)


<!-- If your change includes user interface changes, please add an image, or an animation "An image is worth a thousand words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create animations -->

## Type of change
<!-- Please delete options that are not relevant. -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update (not sure about this)

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->
based on discussions in #3994 and #5024, closes #5024, closes #3994 

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

RobAnKo committed:
add keybindings ([ and ] for label brush size change; add autorepeatable keybindings

RobAnKo committed:
adjust number of Labels methods (12 -> 14)

RobAnKo committed:
clean up ActionManager attributes, add some docstrings

Czaki commented on napari/layers/labels/_labels_key_bindings.py:
```diff
@@ -90,6 +90,26 @@ def increase_label_id(layer: Labels):
+)
+def decrease_brush_size(layer: Labels):
+    """Decrease the brush size"""
+    if layer.brush_size > 1:  # here we should probably add a non-hard-coded
```
move to module level constraint

RobAnKo replied:
Hey @Czaki, thanks for your suggestions, it is way cleaner this way. 
I'll implement them later today.

One question at this point: can you elaborate on what you mean with "move to module level constraint" here?

Czaki replied:
Create variables `MIN_BRUSH_SIZE = 1` and MAX_BRUSH_SIZE = 40` and compare to them, instead of hardcoded variables. 

This is halfway to making it fully configurable because someone could modify them from code.

Czaki commented on napari/layers/labels/_labels_key_bindings.py:
```diff
@@ -90,6 +90,26 @@ def increase_label_id(layer: Labels):
+)
+def increase_brush_size(layer: Labels):
+    """Increase the brush size"""
+    if layer.brush_size < 40:  # here we should probably add a non-hard-coded
```
move to module level containt

Czaki commented on napari/utils/action_manager.py:
```diff
@@ -37,6 +38,7 @@ class Action:
     command: Callable
     description: str
     keymapprovider: KeymapProvider  # subclassclass or instance of a subclass
+    repeatable: bool
```
```suggestion
    repeatable: bool = False
```

Czaki commented on napari/utils/action_manager.py:
```diff
@@ -99,12 +101,34 @@ def _validate_action_name(self, name):
+            return repeatable_actions[name]
+        else:
+            return False
+            # default is non-repeatable
```
```suggestion
        return self.actions[name].repeatable
```

Czaki commented on napari/utils/action_manager.py:
```diff
@@ -150,7 +179,9 @@ def register_action(
         self._validate_action_name(name)
-        self._actions[name] = Action(command, description, keymapprovider)
+        self._actions[name] = Action(
+            command, description, keymapprovider, self._is_repeatable(name)
```
```suggestion
            command, description, keymapprovider, repeatable
```

Czaki commented on napari/utils/action_manager.py:
```diff
@@ -381,6 +412,34 @@ def _get_active_shortcuts(self, active_keymap):
+        list
+            List of shortcuts that are repeatable.
+        """
+        active_func_names = [i[1].__name__ for i in active_keymap.items()]
```
```suggestion
        active_func_names = set(i[1].__name__ for i in active_keymap.items())
```

Czaki commented on napari/utils/action_manager.py:
```diff
@@ -381,6 +412,34 @@ def _get_active_shortcuts(self, active_keymap):
+                and action.repeatable
+            ):
+                for shortcut in shortcuts:
+                    active_repeatable_shortcuts.append(shortcut)
```
```suggestion
                active_repeatable_shortcuts.extend(shortcuts)

```

Czaki commented on napari/utils/key_bindings.py:
```diff
@@ -508,24 +509,35 @@ def release_key(self, key_combo):
+            "Down",
+            "Left",
+            "Right",
+        ]
```
```suggestion
        repeatables = {
            *action_manager._get_repeatable_shortcuts(self.keymap_chain),
            "Up",
            "Down",
            "Left",
            "Right",
        }
```

Czaki commented on napari/utils/repeatable_actions.py:
```diff
@@ -0,0 +1,50 @@
+    'napari:finish_drawing_shape': False,
+    'napari:reset_active_layer_affine': False,
+    'napari:transform_active_layer': False,
+}
```
```suggestion
```

This should be set when registering actions

Czaki commented on napari/utils/action_manager.py:
```diff
@@ -8,6 +8,7 @@
 from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional
 
 from ..utils.events import EmitterGroup
+from ..utils.repeatable_actions import repeatable_actions
```
```suggestion
```

Czaki commented on napari/layers/utils/layer_utils.py:
```diff
@@ -51,6 +51,7 @@ def _inner(func):
             command=func,
             description=description,
             keymapprovider=keymapprovider,
+            repeatable=False,
```
Should be a passed argument of `register_layer_action` that needs to be added.

Czaki reviewed (commented):
I'm happy to see that this thing is changing. Maybe even @andy-sweet will allow us to push it to `rc1`.  

If we create some structure to check if a member belongs to it, we should use `set` because of performance. 

I prefer to store information if an action is repeatable next to its definition, not in a separate structure.

RobAnKo committed:
Apply suggestions from code review - update of register_layer_action still pending

- use set instead of list for active_func_names
- remove external structure repeatable_actions (to do instead: store repeatability with each action, defined at time of registration, by adding argument 'repeatable' to register_layer_action)
- clean function definitions

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki commented on napari/layers/labels/_labels_key_bindings.py:
```diff
@@ -90,6 +90,26 @@ def increase_label_id(layer: Labels):
 
+@register_label_action(
+    trans._("Decrease the paint brush size by one."),
+)
```
```suggestion
@register_label_action(
    trans._("Decrease the paint brush size by one."), repetable=True
)
```

Czaki commented on napari/layers/labels/_labels_key_bindings.py:
```diff
@@ -90,6 +90,26 @@ def increase_label_id(layer: Labels):
+
+@register_label_action(
+    trans._("Increase the paint brush size by one."),
+)
```
```suggestion
@register_label_action(
    trans._("Increase the paint brush size by one."), repetable=True
)
```

RobAnKo committed:
add 'repeatable' to register_layer_action, adjust layer-specific wrapper functions, set defaults to non-repeatable

RobAnKo committed:
Merge branch 'napari:main' into autorepeatable-keybindings

Czaki reviewed (approved):
Looks nice and ready to merge. Thanks for your contribution.

jni reviewed (approved):
Hahahaha that is a *badass* screengrab @RobAnKo! I was grinning ear-to-ear watching it. 😂 Thanks for everyone who has participated so far. Just want to lend my support to landing this in 0.4.17. 👍 🎉 🚀

andy-sweet commented on napari/utils/action_manager.py:
```diff
@@ -134,6 +155,10 @@ def register_action(
             registered. This make sure the shortcut is active only when an
             instance of this is in focus.
+        repeatable : bool
+            a boolean flag indictation whether the action can be autorepeated.
```
```suggestion
            a boolean flag indicating whether the action can be autorepeated.
```

andy-sweet commented on napari/utils/action_manager.py:
```diff
@@ -99,12 +100,30 @@ def _validate_action_name(self, name):
 
+    def _is_repeatable(self, name) -> bool:
+        """
+
```
Nit: remove the whitespace. Or maybe better, just condense the whole docstring into a one or two line summary as this is a private/helper function that is relatively simple.

Actually, is this used? I did a quick search for it on the GitHub file diff and did not find it. If that's correct, just delete this as the public equivalent is just as readable as calling this method (i.e. `self.actions[name].repeatable` vs.` self._is_repeatable(name)`).

RobAnKo replied:
Thanks for the nit-picks, I'll change them right away.

Good call on this private function, defined it during development and failed to realize it's not needed in the current implementation.

RobAnKo committed:
Merge remote-tracking branch 'upstream/main' into autorepeatable-keybindings

RobAnKo committed:
apply suggestions: remove _is_repeatable; fix docstring typo
